File: JavaScript/src/Internals/Functions/MathBase/Function_ROUNDUP.js
import { Function_2 } from '../Function_2.js';
import { Operand } from '../../../Operand.js';

class Function_ROUNDUP extends Function_2 {
    get Name() {
        return "RoundUp";
    }

    constructor(z) {
    super(z);
  }

    evaluate(engine, tempParameter) {
        let args1 = this.getNumber_1(engine, tempParameter);
        if (args1.IsError) { return args1; }
        let args2 = this.getNumber_2(engine, tempParameter);
        if (args2.IsError) { return args2; }
        if (args1.NumberValue == 0.0) { return args1; }
        let a = Math.pow(10, args2.IntValue);
        let b = args1.NumberValue;

        let t = (Math.ceil(Math.abs(b) * a)) / a;
        if (b > 0) return Operand.Create(t);
        return Operand.Create(-t);
    }
}

export { Function_ROUNDUP };

File: java/toolgood.algorithm/src/main/java/toolgood/algorithm/internals/functions/mathBase/Function_ROUNDUP.java
package toolgood.algorithm.internals.functions.mathBase;

import java.math.BigDecimal;
import java.util.List;
import java.util.function.BiFunction;
import toolgood.algorithm.AlgorithmEngine;
import toolgood.algorithm.Operand;
import toolgood.algorithm.enums.OperandType;
import toolgood.algorithm.internals.functions.NoneEngine;
import toolgood.algorithm.internals.ParameterType;
import toolgood.algorithm.internals.functions.FunctionBase;
import toolgood.algorithm.internals.functions.Function_2;

public final class Function_ROUNDUP extends Function_2 {
    public Function_ROUNDUP(FunctionBase[] funcs) {
        super(funcs);
        if (funcs.length != 2) {
            throw new IllegalArgumentException("Function '" + Name() + "' requires exactly 2 parameters.");
        }
    }

    @Override
    public String Name() {
        return "RoundUp";
    }

    @Override
    public Operand Evaluate(AlgorithmEngine engine, BiFunction<AlgorithmEngine, String, Operand> tempParameter) {
        Operand args1 = GetNumber_1(engine, tempParameter);
        if (args1.IsErrorOrNone()) {
            return args1;
        }
        Operand args2 = GetNumber_2(engine, tempParameter);
        if (args2.IsErrorOrNone()) {
            return args2;
        }
        if (args2.IntValue() < -15 || args2.IntValue() > 15) {
            return ParameterError(2);
        }
        if (args1.NumberValue().compareTo(BigDecimal.ZERO) == 0) {
            return args1;
        }
        BigDecimal b = args1.NumberValue();
        int digits = args2.IntValue();
        double a = Math.pow(10, digits);
        double t = Math.ceil(Math.abs(b.doubleValue()) * a) / a;
        if (b.compareTo(BigDecimal.ZERO) > 0) {
            return Operand.Create(BigDecimal.valueOf(t));
        }
        return Operand.Create(BigDecimal.valueOf(-t));
    }

    @Override
    public OperandType GetResultType() {
        return OperandType.NUMBER;
    }

    @Override
    public void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType, String op, String val) {
        func1.GetParameterTypes(noneEngine, result, OperandType.NUMBER);
        func2.GetParameterTypes(noneEngine, result, OperandType.NUMBER);
    }
}
